### DISCUSSION TITLE
Does (Open) Watcom have an H2INC tool?

### DISCUSSION THREAD

**Discussion by volkertb** [open-watcom/open-watcom-v2#731] [Ideas]
Hi,

I'm trying to "port" [SoftMPU](https://github.com/bjt42/softmpu) to build with the Open Watcom v2 toolchain (C compiler and WASM) instead of Microsoft Visual C and MASM. One of the steps in the initial Microsoft-specific build process is the invocation of a tool called `H2INC`, which is apparently part of MASM, and can convert C header files into assembly .INC files, which allows C code to be combined with assembly code.

I've looked for (and found) some alternative open-source H2INC replacements or workalikes, but since those aren't Watcom-specific, they might not produce .INC files that are compatible with WASM (in terms of assembly language dialect).

I couldn't find a similar tool as part of the Open Watcom v2 distribution. But perhaps such functionality is hidden in one the utilities that are included in Open Watcom (v2), perhaps behind an optional command-line parameter? Or maybe there's a simpler way to convert C header (.h) files to .INC files that are compatible with WASM?

Thanks.

**Comment by jmalak** [comment]
If I remember properly then OW has only such tool for Fortran source code.
You could be able to use Microsoft tool for converting .h to .inc files.
Try to use old version from MASM 5.

**Comment by volkertb** [comment]
I appreciate the suggestion, but the whole point of my exercise here is to make SoftMPU independent of the Microsoft toolchain, or any closed-source tools. The existing build script (a BAT file) actually already uses Microsoft's H2INC tool (the one distributed with MASM 6).

Anyway, I did find some open source alternatives, that I am going to try. For reference to anyone else stumbling upon this GitHub issue thread:

- [h2incx](https://masm32.com/board/index.php?topic=7006.msg75149#msg75149) by Japheth a.k.a. @Baron-von-Riedesel (which I backed up at https://github.com/volkertb/h2incx)
- https://github.com/Lerking/h2inc
- https://github.com/Lerking/h2inc-old
- https://github.com/JonathonReinhart/h2incn (generates NASM compatible include files, but I hope it won't be too hard to make those work with WASM too)

If anybody here has any other suggestions, please let me know! Thanks. :slightly_smiling_face:

**Comment by volkertb** [comment]
@jmalak By the way, if you don't plan on including any such tool in Open Watcom v2, feel free to convert this issue into a discussion. Thanks.
